Test de-flake

https://github.com/ankidroid/Anki-Android/actions/runs/24023107089/job/70055885488

```
PreviewerViewModelTest > backSideOnly with navigation FAILED
    java.lang.AssertionError: Should show answer
        at org.junit.Assert.fail(Assert.java:89)
        at org.junit.Assert.assertTrue(Assert.java:42)
        at com.ichi2.anki.previewer.PreviewerViewModelTest$backSideOnly with navigation$1.invokeSuspend(PreviewerViewModelTest.kt:250)
        at com.ichi2.anki.previewer.PreviewerViewModelTest$backSideOnly with navigation$1.invoke(PreviewerViewModelTest.kt)
        at com.ichi2.anki.previewer.PreviewerViewModelTest$backSideOnly with navigation$1.invoke(PreviewerViewModelTest.kt)
        at com.ichi2.anki.libanki.testutils.AnkiTest.runTestInner$suspendImpl(AnkiTest.kt:375)
        at com.ichi2.anki.libanki.testutils.AnkiTest.runTestInner(AnkiTest.kt)
        at com.ichi2.anki.libanki.testutils.InMemoryAnkiTest.runTestInner(InMemoryAnkiTest.kt:36)
        at com.ichi2.anki.libanki.testutils.AnkiTest$runTest$1$1.invokeSuspend(AnkiTest.kt:368)
        at com.ichi2.anki.libanki.testutils.AnkiTest$runTest$1$1.invoke(AnkiTest.kt)
        at com.ichi2.anki.libanki.testutils.AnkiTest$runTest$1$1.invoke(AnkiTest.kt)
        at kotlinx.coroutines.test.TestBuildersKt__TestBuildersKt$runTest$2$1$1.invokeSuspend(TestBuilders.kt:317)
        at kotlin.coroutines.jvm.internal.BaseContinuationImpl.resumeWith(ContinuationImpl.kt:34)
        at kotlinx.coroutines.DispatchedTask.run(DispatchedTask.kt:100)
        at kotlinx.coroutines.test.TestDispatcher.processEvent$kotlinx_coroutines_test(TestDispatcher.kt:24)
        at kotlinx.coroutines.test.TestCoroutineScheduler.tryRunNextTaskUnless$kotlinx_coroutines_test(TestCoroutineScheduler.kt:99)
        at kotlinx.coroutines.test.TestBuildersKt__TestBuildersKt$runTest$2$1$workRunner$1.invokeSuspend(TestBuilders.kt:326)
        at kotlin.coroutines.jvm.internal.BaseContinuationImpl.resumeWith(ContinuationImpl.kt:34)
        at kotlinx.coroutines.DispatchedTask.run(DispatchedTask.kt:100)
        at kotlinx.coroutines.EventLoopImplBase.processNextEvent(EventLoop.common.kt:263)
        at kotlinx.coroutines.BlockingCoroutine.joinBlocking(Builders.kt:94)
        at kotlinx.coroutines.BuildersKt__BuildersKt.runBlocking(Builders.kt:70)
        at kotlinx.coroutines.BuildersKt.runBlocking(Unknown Source)
        at kotlinx.coroutines.BuildersKt__BuildersKt.runBlocking$default(Builders.kt:48)
        at kotlinx.coroutines.BuildersKt.runBlocking$default(Unknown Source)
        at kotlinx.coroutines.test.TestBuildersJvmKt.createTestResult(TestBuildersJvm.kt:10)
        at kotlinx.coroutines.test.TestBuildersKt__TestBuildersKt.runTest-8Mi8wO0(TestBuilders.kt:309)
        at kotlinx.coroutines.test.TestBuildersKt.runTest-8Mi8wO0(Unknown Source)
        at kotlinx.coroutines.test.TestBuildersKt__TestBuildersKt.runTest-8Mi8wO0(TestBuilders.kt:167)
        at kotlinx.coroutines.test.TestBuildersKt.runTest-8Mi8wO0(Unknown Source)
        at com.ichi2.anki.libanki.testutils.AnkiTest.runTest-dWUq8MI(AnkiTest.kt:367)
        at com.ichi2.anki.libanki.testutils.InMemoryAnkiTest.runTest-dWUq8MI(InMemoryAnkiTest.kt:36)
        at com.ichi2.anki.libanki.testutils.AnkiTest.runTest-dWUq8MI$default(AnkiTest.kt:356)
        at com.ichi2.anki.previewer.PreviewerViewModelTest.backSideOnly with navigation(PreviewerViewModelTest.kt:217)
```

--- a/AnkiDroid/src/main/java/com/ichi2/anki/previewer/PreviewerViewModel.kt
+++ b/AnkiDroid/src/main/java/com/ichi2/anki/previewer/PreviewerViewModel.kt
@@ -38,7 +38,6 @@ import com.ichi2.anki.utils.ext.require
 import com.ichi2.anki.utils.ext.setUserFlagForCards
 import kotlinx.coroutines.Deferred
 import kotlinx.coroutines.flow.MutableStateFlow
-import kotlinx.coroutines.flow.collectLatest
 import kotlinx.coroutines.flow.combine
 import kotlinx.coroutines.flow.update
 import timber.log.Timber
@@ -88,8 +87,8 @@ class PreviewerViewModel(
     @NeedsTest("16302 - a sound-only card on the back/flipped with 'don't keep activities'")
     @NeedsTest("16302 - on config changes, sound continues to play")
     override fun onPageFinished(isAfterRecreation: Boolean) {
-        if (isAfterRecreation) {
-            launchCatchingIO {
+        launchCatchingIO {
+            if (isAfterRecreation) {
                 showCard(showAnswerOnReload)
                 // isAfterRecreation can either mean:
                 // * after config change (ViewModel exists)
@@ -97,13 +96,9 @@ class PreviewerViewModel(
                 // if the ViewModel existed, we want to continue playing audio
                 // if not, we want to setup the sound player
                 cardMediaPlayer.ensureAvTagsLoaded(currentCard.await())
-            }
-            return
-        }
-        launchCatchingIO {
-            currentIndex.collectLatest {
-                showCard(showAnswer = backSideOnly.value)
-                loadAndPlaySounds()
+            } else {
+                // re-render the current card
+                updateCurrentIndex { it }
             }
         }
     }
@@ -159,7 +154,7 @@ class PreviewerViewModel(
                 showAnswer()
                 cardMediaPlayer.autoplayAllForSide(CardSide.ANSWER)
             } else {
-                currentIndex.update { it + 1 }
+                updateCurrentIndex { it + 1 }
             }
         }
     }
@@ -171,7 +166,7 @@ class PreviewerViewModel(
     fun onPreviousButtonClick() {
         launchCatchingIO {
             if (currentIndex.value > 0) {
-                currentIndex.update { it - 1 }
+                updateCurrentIndex { it - 1 }
             } else if (showingAnswer.value && !backSideOnly.value) {
                 showQuestion()
             }
@@ -196,13 +191,20 @@ class PreviewerViewModel(
         val index = sliderPosition - 1
         if (index !in selectedCardIds.indices) return
         launchCatchingIO {
-            currentIndex.emit(index)
+            updateCurrentIndex { index }
         }
     }
 
     /* *********************************************************************************************
      *************************************** Internal methods ***************************************
      ********************************************************************************************* */
+
+    /** Applies [update] to [currentIndex] and re-renders the resulting card. */
+    private suspend fun updateCurrentIndex(update: (Int) -> Int) {
+        currentIndex.update(update)
+        showCard(showAnswer = backSideOnly.value)
+        loadAndPlaySounds()
+    }
 
     private suspend fun showCard(showAnswer: Boolean) {
         currentCard =

--- a/AnkiDroid/src/test/java/com/ichi2/anki/previewer/PreviewerViewModelTest.kt
+++ b/AnkiDroid/src/test/java/com/ichi2/anki/previewer/PreviewerViewModelTest.kt
@@ -22,8 +22,6 @@ import com.ichi2.anki.browser.IdsFile
 import com.ichi2.anki.servicelayer.NoteService
 import com.ichi2.anki.utils.ext.flag
 import com.ichi2.testutils.JvmTest
-import com.ichi2.testutils.common.Flaky
-import com.ichi2.testutils.common.OS
 import io.mockk.coEvery
 import io.mockk.every
 import io.mockk.mockk
@@ -103,7 +101,6 @@ class PreviewerViewModelTest : JvmTest() {
         }
 
     @Test
-    @Flaky(OS.ALL)
     fun `previous button`() =
         runTest {
             // Start at Index 1
@@ -122,7 +119,6 @@ class PreviewerViewModelTest : JvmTest() {
         }
 
     @Test
-    @Flaky(OS.ALL)
     fun `toggle back side only`() =
         runTest {
             assertFalse(viewModel.backSideOnly.value) // initial state should be false

